src/Microsoft.ServiceHub.Framework/SystemTextJsonConverters.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.ServiceHub.Framework;
+
+/// <summary>
+/// Contains custom converters to workaround <see href="https://github.com/dotnet/runtime/issues/85981">this bug in System.Text.Json</see>.
+/// </summary>
+/// <remarks>
+/// We should be able to remove this class once we upgrade to System.Text.Json 8.0.0.
+/// </remarks>
+internal static class SystemTextJsonConverters
+{
+	/// <summary>
+	/// A custom <see cref="Version"/> converter.
+	/// </summary>
+	internal class VersionConverter : JsonConverter<Version>
+	{
+		/// <summary>
+		/// The singleton instance of this converter.
+		/// </summary>
+		internal static readonly VersionConverter Instance = new VersionConverter();
+
+		private VersionConverter()
+		{
+		}
+
+		/// <inheritdoc/>
+		public override Version Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+		{
+			try
+			{
+				return Version.Parse(reader.GetString()!);
```

Comment:
NB you don't need to reimplement the converter from scratch, just wrapping the default should be good enough:
```C#
public sealed class JsonConverterWrapper<T> : JsonConverter<T>
{
    private readonly JsonConverter<T> _defaultConverter = (JsonConverter<T>)JsonSerializerOptions.Default.GetTypeInfo(typeof(T));

    public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
        => _defaultConverter.Write(writer, value, options);

    public override T? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => _defaultConverter.Read(ref reader, typeToConvert, options);

    public override T ReadAsPropertyName(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => _defaultConverter.ReadAsPropertyName(ref reader, typeToConvert, options);

    public override void WriteAsPropertyName(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
        => _defaultConverter.WriteAsPropertyName(writer, value, options);
}
```
STJ will detect that it's a user-defined converter and run all validation steps including checking for null.

Reply:
I love it. Thanks.